### PR TITLE
Support C++20

### LINKED ISSUE #5993: Support C++20

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #5993 

Problem Summary:

### What is changed and how it works?

Thanks to @fuzhe1989 previous excellent work, we can support C++20 now. Check https://en.cppreference.com/w/cpp/20 and https://www.geeksforgeeks.org/features-of-c-20/, to get familiar with new features of C++20.

This pr:

1. `set (CMAKE_CXX_STANDARD 20)`
2. update `go1.17` to `go1.20`
3. rewrite some code with C++20

TODO:
- [ ] update to clang-15

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support C++20
```

### CODE REVIEW

File: release-centos7-llvm/dockerfiles/misc/bake_llvm_base_amd64.sh [hunk 1]
```diff
@@ -33,7 +33,7 @@ function bake_llvm_base_amd64() {
 
     # LLVM
     source $SCRIPTPATH/bootstrap_llvm.sh
-    bootstrap_llvm "13.0.0"
```
Comment (breezewish) [inline release-centos7-llvm/dockerfiles/misc/bake_llvm_base_amd64.sh:36]: Could we use (latest minor fix version) LLVM 15.0.7? It should contain some useful bug fixes.

File: dbms/src/Flash/Mpp/MPPTunnel.cpp [hunk 2]
```diff
@@ -208,7 +208,7 @@ void MPPTunnel::connectLocalV2(size_t source_index, LocalRequestHandler & local_
 {
     {
         std::unique_lock lk(mu);
-        RUNTIME_CHECK_MSG(status == TunnelStatus::Unconnected, fmt::format("MPPTunnel {} has connected or finished: {}", tunnel_id, statusToString()));
+        RUNTIME_CHECK_MSG(status == TunnelStatus::Unconnected, fmt::runtime(fmt::format("MPPTunnel {} has connected or finished: {}", tunnel_id, statusToString())));
```
Comment (breezewish) [inline dbms/src/Flash/Mpp/MPPTunnel.cpp:211]:
```suggestion
        RUNTIME_CHECK_MSG(status == TunnelStatus::Unconnected, "MPPTunnel {} has connected or finished: {}", tunnel_id, statusToString());
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlock.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <fmt/core.h>
```
Comment (breezewish) [inline dbms/src/Flash/Coprocessor/DAGQueryBlock.cpp:15]: Useless include?

**Review by breezewish — APPROVED** [review]
The rest LGTM

File: release-centos7-llvm/dockerfiles/misc/bake_llvm_base_amd64.sh (hunk 1 above)
Comment (Lloyd-Pottiger) [reply release-centos7-llvm/dockerfiles/misc/bake_llvm_base_amd64.sh:36]: yes, updated.

File: dbms/src/Flash/Coprocessor/DAGQueryBlock.cpp (hunk 3 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Flash/Coprocessor/DAGQueryBlock.cpp]: remove.

File: dbms/src/Flash/Mpp/MPPTunnel.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Flash/Mpp/MPPTunnel.cpp]: fixed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -159,7 +159,7 @@ endif ()
 
 include (cmake/cpu_features.cmake)
 
-set (CMAKE_CXX_STANDARD 17)
+set (CMAKE_CXX_STANDARD 20)
```
Comment (solotzg) [inline CMakeLists.txt:162]: Can it work under MacOS x86|arm64 as well?

Comment (Lloyd-Pottiger) [reply CMakeLists.txt:162]: Work under MacOS arm64.

File: cmake/cpu_features.cmake [hunk 5]
```diff
@@ -165,8 +165,9 @@ elseif (ARCH_AMD64)
     set (TEST_FLAG "-mpopcnt")
     set (CMAKE_REQUIRED_FLAGS "${TEST_FLAG} -O0")
     check_cxx_source_compiles("
+        #include <bit>
         int main() {
-            auto a = __builtin_popcountll(0);
+            auto a = std::popcount(0);
```
Comment (solotzg) [inline cmake/cpu_features.cmake:170]:
```
candidate template ignored: substitution failure [with _Tp = int]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
popcount(_Tp __t) noexcept
^
1 error generated.
```

```suggestion
            auto a = std::popcount(0u);
```

File: cmake/cpu_features.cmake [hunk 6]
```diff
@@ -165,8 +165,9 @@ elseif (ARCH_AMD64)
...
             (void)a;
             return 0;
         }
```
Comment (solotzg) [inline cmake/cpu_features.cmake:173]: Compiler will generate errors.

Comment (solotzg) [reply cmake/cpu_features.cmake:173]: - `std::popcount(0);` --> `std::popcount(0u); `
- need flag like `-std=gnu++20`

Comment (Lloyd-Pottiger) [reply cmake/cpu_features.cmake]: reverted.

File: dbms/src/Flash/Mpp/MPPTunnelSet.cpp [hunk 7]
```diff
@@ -29,7 +29,7 @@ namespace
 void checkPacketSize(size_t size)
 {
     static constexpr size_t max_packet_size = 1u << 31;
-    RUNTIME_CHECK(size < max_packet_size, fmt::format("Packet is too large to send, size : {}", size));
+    RUNTIME_CHECK_MSG(size < max_packet_size, fmt::runtime(fmt::format("Packet is too large to send, size : {}", size)));
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Mpp/MPPTunnelSet.cpp:32]:
```suggestion
    RUNTIME_CHECK_MSG(size < max_packet_size, "Packet is too large to send, size : {}", size);
```

**Comment by Lloyd-Pottiger** [comment]
Meet some errors while compiling llvm project, revert update to clang-15, open another pr after solving this problem.

**Pull request merged**